src/Middleware/Localization/src/RequestLocalizationMiddleware.cs

```diff
@@ -162,35 +161,68 @@ private static void SetCurrentThreadCulture(RequestCulture requestCulture)
         }
 
         private static CultureInfo GetCultureInfo(
-            IList<StringSegment> cultureNames,
-            IList<CultureInfo> supportedCultures,
+            IEnumerable<StringSegment> cultureNames,
+            IEnumerable<CultureInfo> supportedCultures,
             bool fallbackToParentCultures)
         {
-            foreach (var cultureName in cultureNames)
+            var nonNullableCultureNames = cultureNames.Where(c => c != null);
+
+            foreach (var cultureName in nonNullableCultureNames)
+            {
+                var cultureInfo = GetCultureInfo(cultureName, supportedCultures, fallbackToParentCultures);
+                if (cultureInfo != null)
+                {
+                    return cultureInfo;
+                }
+            }
+
+            return null;
+        }
+
+        private static CultureInfo GetCultureInfo(
+            StringSegment cultureName,
+            IEnumerable<CultureInfo> supportedCultures,
+            bool fallbackToParentCultures)
+        {
+            var culture = GetCultureInfo(cultureName, supportedCultures);
+
+            if (culture == null)
+            {
+                return culture;
+            }
+
+            if (fallbackToParentCultures)
             {
-                // Allow empty string values as they map to InvariantCulture, whereas null culture values will throw in
-                // the CultureInfo ctor
-                if (cultureName != null)
+                var fallbackCulture = culture;
+
+                while (fallbackCulture != fallbackCulture.Parent)
```

Comment:
I don't think this does what you think it does. The call to `GetCultureInfo(StringSegment, IEnumerable<CultureInfo>)` on line 187 will return `null` (`default(CultureInfo)`) if the cultureName StringSegment is not equal to the precise name of one of the SupportedCultures. If that DOES return a non-Null value then that value is already a supported culture and we don't need to fall back at all.

In order to overcome that you would have to create a new `CultureInfo` based on the user input, which is dangerously expensive, and avoiding that scenario is how we ended up with the current code in the first place.

Reply:
I will revise the code again, may be I need remove some of the existing stuff, don't blame me I'm still working with the old localization repo, because I'm still unable to develop on the new code base after the migration

> In order to overcome that you would have to create a new CultureInfo based on the user input, which is dangerously expensive, and avoiding that scenario is how we ended up with the current code in the first place.

I'm trying to avoid that too, my aim is to fallback whenever it's possible

BTW @ryanbrandenburg I need your answer about my comment here https://github.com/aspnet/AspNetCore/pull/14964#issuecomment-541428814

Reply:
I will revise the code again, may be I need remove some of the existing stuff, don't blame me I'm still working with the old localization repo, because I'm still unable to develop on the new code base after the migration

> In order to overcome that you would have to create a new CultureInfo based on the user input, which is dangerously expensive, and avoiding that scenario is how we ended up with the current code in the first place.

I'm trying to avoid that too, my aim is to fallback whenever it's possible

BTW @ryanbrandenburg I need your answer about my comment here https://github.com/aspnet/AspNetCore/pull/14964#issuecomment-541428814